package com.adverto.dejonghe.server.views.subViews.toolsSubView;

import com.adverto.dejonghe.common.dbservices.ProductService;
import com.adverto.dejonghe.common.entities.customers.Customer;
import com.adverto.dejonghe.common.entities.enums.workorder.Tools;
import com.adverto.dejonghe.common.entities.product.product.Product;
import com.adverto.dejonghe.server.customEvents.AddRemoveProductEvent;
import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.button.ButtonVariant;
import com.vaadin.flow.component.dialog.Dialog;
import com.vaadin.flow.component.html.H3;
import com.vaadin.flow.component.html.Span;
import com.vaadin.flow.component.icon.VaadinIcon;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.component.textfield.TextField;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;

@Component
@Scope("prototype")
public class ToolsFuelView extends VerticalLayout {

    ProductService productService;
    ApplicationEventPublisher eventPublisher;

    H3 title;
    Tools selectedTool;
    List<Product> selectedProducts;
    Integer amountFuel = 1;
    TextField tfFuel;
    Integer selectedTeam;
    Boolean bAgro;

    @Autowired
    public void ToolsFuelView(ProductService productService,
                              ApplicationEventPublisher eventPublisher) {
        this.productService = productService;
        this.eventPublisher = eventPublisher;

        title = new H3();
        this.setAlignItems(Alignment.CENTER);

        Button okButton = new Button("Voeg toe");
        setUpOkButton(okButton);
        this.add(title);
        add(getFuelComponent());
        this.add(okButton);
    }

    private HorizontalLayout getFuelComponent() {
        HorizontalLayout horizontalLayout = new HorizontalLayout();
        horizontalLayout.setSpacing(true);
        tfFuel = new TextField();
        tfFuel.addValueChangeListener(event -> {
            String value = event.getValue();

            if (value == null || value.isBlank()) {
                amountFuel = 0;
                return;
            }

            try {
                amountFuel = Integer.parseInt(value.replace(",", "."));
            } catch (NumberFormatException e) {
                // Ongeldige invoer
                amountFuel = 0;
            }
        });
        Button minusButton = new Button(VaadinIcon.MINUS.create());
        minusButton.addClickListener(buttonClickEvent ->{
            amountFuel--;
            tfFuel.setValue(amountFuel.toString());
                });
        Button plusButton = new Button(VaadinIcon.PLUS.create());
        plusButton.addClickListener(buttonClickEvent ->{
            amountFuel++;
            tfFuel.setValue(amountFuel.toString());
        });

        tfFuel.setSuffixComponent(new Span("liter brandstof"));
        tfFuel.setValue(amountFuel.toString());

        horizontalLayout.add(minusButton, tfFuel,  plusButton);
        return horizontalLayout;
    }

    private void setUpOkButton(Button okButton) {
        okButton.addThemeVariants(ButtonVariant.LUMO_SUCCESS);
        okButton.addClickListener(e -> {
            Product productToAdd = productService.findByProductCodeContaining("OPVER-brandstof").get().get(0);
            productToAdd.setTeamNumber(selectedTeam);
            productToAdd.setSelectedAmount(Double.valueOf(amountFuel));
            productToAdd.setInternalName("Brandstof : " + selectedTool.getDiscription());
            if(!bAgro) {
                if(productToAdd.getSellPriceIndustry() == 0.0){
                    productToAdd.setTotalPrice(Double.valueOf(amountFuel) * productToAdd.getSellPrice());
                }
                else{
                    productToAdd.setTotalPrice(Double.valueOf(amountFuel) * productToAdd.getSellPriceIndustry());
                }
            }
            else{
                productToAdd.setTotalPrice(Double.valueOf(amountFuel) * productToAdd.getSellPrice());
            }
            productToAdd.setAbbreviation(selectedTool.getAbbreviation());
            selectedProducts.add(productToAdd);

            Product productToAdd2 = productService.findByProductCodeContaining(selectedTool.getAbbreviation()).get().get(0);
            productToAdd2.setSelectedAmount(1.0);
            productToAdd2.setTeamNumber(selectedTeam);
            productToAdd2.setInternalName("Forfait : " + selectedTool.getDiscription());
            if(!bAgro) {
                if(productToAdd.getSellPriceIndustry() == 0.0){
                    productToAdd2.setTotalPrice(1.0 * productToAdd2.getSellPrice());
                }
                else{
                    productToAdd2.setTotalPrice(1.0 * productToAdd2.getSellPriceIndustry());
                }
            }
            else{
                productToAdd2.setTotalPrice(1.0 * productToAdd2.getSellPrice());
            }
            productToAdd2.setAbbreviation(selectedTool.getAbbreviation());
            selectedProducts.add(productToAdd2);
            eventPublisher.publishEvent(new AddRemoveProductEvent(this, "Product toegevoegd",null));

            getParent().ifPresent(parent -> {
                if (parent instanceof Dialog dialog) {
                    dialog.close();
                }
            });

        });
    }

    public Tools getSelectedTool() {
        return selectedTool;
    }

    public void setSelectedToolTeam(Tools selectedTool, Integer selectedTeam) {
        amountFuel = 1;
        tfFuel.setValue(amountFuel.toString());
        title.setText(selectedTool.getDiscription() + " selecteren?");
        this.selectedTool = selectedTool;
        this.selectedTeam = selectedTeam;
    }

    public List<Product> getSelectedProducts() {
        return selectedProducts;
    }

    public void setSelectedProducts(List<Product> selectedProducts) {
        this.selectedProducts = selectedProducts;
    }

    public void setCustomerByWorkAddress(Optional<List<Customer>> customerByWorkAddress) {
        try{
            if(customerByWorkAddress.get().getFirst().getBIndustry()){
                bAgro = !customerByWorkAddress.get().getFirst().getBIndustry();
            }
            else{
                bAgro = true;
            }
        }
        catch(Exception e){
            bAgro = true;
        }

    }
}
